File: src/minigames/apocalypse_survival/java/work/lclpnet/ap2/game/apocalypse_survival/goal/RoamGoal.java
package work.lclpnet.ap2.game.apocalypse_survival.goal;

import net.minecraft.world.entity.PathfinderMob;
import net.minecraft.world.entity.ai.goal.Goal;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.game.apocalypse_survival.util.TargetManager;

import java.util.EnumSet;

public class RoamGoal extends Goal {

    private final PathfinderMob mob;
    private final TargetManager targetManager;
    private final double speed;
    private @Nullable Vec3 target = null;

    public RoamGoal(PathfinderMob mob, TargetManager targetManager, double speed) {
        this.mob = mob;
        this.targetManager = targetManager;
        this.speed = speed;

        setFlags(EnumSet.of(Flag.MOVE));
    }

    @Override
    public boolean canUse() {
        if (mob.hasControllingPassenger() || mob.getTarget() != null || mob.getNavigation().isInProgress()) {
            return false;
        }

        target = targetManager.getDensityManager().startGuarding(mob);

        return target != null;
    }

    @Override
    public boolean canContinueToUse() {
        return !mob.getNavigation().isDone() && !mob.hasControllingPassenger();

    }

    @Override
    public void start() {
        if (target == null) return;

        mob.getNavigation().moveTo(target.x(), target.y(), target.z(), speed);
    }

    @Override
    public void stop() {
        mob.getNavigation().stop();
        targetManager.getDensityManager().stopGuarding(mob);
        target = null;
    }
}

File: src/minigames/apocalypse_survival/src/main/kotlin/work/lclpnet/ap2/game/apocalypse_survival/goal/RoamGoal.kt
package work.lclpnet.ap2.game.apocalypse_survival.goal

import net.minecraft.world.entity.PathfinderMob
import net.minecraft.world.entity.ai.goal.Goal
import net.minecraft.world.phys.Vec3
import work.lclpnet.ap2.game.apocalypse_survival.util.TargetManager
import java.util.EnumSet

class RoamGoal(
    private val mob: PathfinderMob,
    private val targetManager: TargetManager,
    private val speed: Double
) : Goal() {

    private var target: Vec3? = null

    init {
        flags = EnumSet.of(Flag.MOVE)
    }

    override fun canUse(): Boolean {
        if (mob.hasControllingPassenger() || mob.target != null || mob.navigation.isInProgress) {
            return false
        }

        target = targetManager.densityManager.startGuarding(mob)

        return target != null
    }

    override fun canContinueToUse() = !mob.navigation.isDone && !mob.hasControllingPassenger()

    override fun start() {
        val t = target ?: return
        mob.navigation.moveTo(t.x(), t.y(), t.z(), speed)
    }

    override fun stop() {
        mob.navigation.stop()
        targetManager.densityManager.stopGuarding(mob)
        target = null
    }
}
